### ISSUE TITLE
Refactor current parser and regex

### ISSUE THREAD

**Issue by schoettl** [200ok-ch/organice#150]
After I added some regexes with `|` to the end of the big regex `markupAndCookieRegex` in #147, we would like to refactor some things in `parse_org.js`:

- [ ] ~split up the monster regex in smaller pieces~ not simply possible as this would create a mess of state-ful regexes and parallel calls to `exec()` on each...
- [ ] ~refactor `parseMarkupAndCookies`: maybe use [`matchAll()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll) instead of `exec()` instead of the while loop~ no, `matchAll()` does not provide the non-matches (normal text inbetween inline-markup)
- [x] refactor `parseMarkupAndCookies`: extract function that creates `lineParts`
- [x] use switch-case instead of repeating else-if constructs

**Comment by schoettl** [comment]
Just a few thoughts: Im not sure if it makes sense to split up the monster regex in smaller regexes. That would mean we have multiple regexes and call multiple `exec()` on them while working on the same input string. That feels like a lot of hacking and not best practice for parsing...

After I wrote the other parser using pegjs yesterday, I'm thinking, it might be better to introduce a small pegjs parser instead of the monster regex. With that we could also fix #157

On the other hand, this work would maybe better invested in the new parser...

**Comment by munen** [comment]
I agree that introducing pegjs for parsing Org files would compromise our workforce. There's already two competing parsers: The currently used one and the new proper parser.

The strategy we're using is this:

1. Keep improving the old one for new features and make bugfixes as long as the new one isn't ready.
2. In parallel, when there's time, work on the new one until there's feature parity with the old one from organice.
3. When the new one is finished, integrate it into organice.

Steps 2 and 3 each will take their time. In the meantime I worked a lot on step 1. Compared to org-web, organices' parser has loads more tests and bugfixes. It also has a few new features. In fact, my initial goal when contributing to org-web was making the parser more robust and bug-free. Due to different goals, the org-web maintainer asked me to do a fork, so I did. And then I spent quite some time improving the parser and making certain that it doesn't create any diff to an existing Org file. That's one of the main features of organice compared to alternative products like Beorg and Orgzly for me. Because I don't want to see unrelated diffs. Those will not inspire confidence in using the tool. I think that's in part why organice is gaining the kind of traction that it does.

Working on step 1 is pragmatic. In a perfect world with more resources, we'd quickly finish steps 2 and 3, of course. On the other hand, the current parser with it's test coverage isn't the worst piece of software in the world, too.

TL;DR: Having said all that, I understand what you're saying, too. Spending too much time refactoring it doesn't make sense, of course. It would be better spent on the new parser. I'm only thinking that we shouldn't make the code of the old parser worse in the meantime. Because that's the one that's in use by loads of people.
